src/Counter.php

```diff
@@ -68,42 +70,68 @@ public function __construct(
      */
     public function hit(string $id): CounterState
     {
-        // Last increment time.
-        // In GCRA it's known as arrival time.
-        $lastIncrementTimeInMilliseconds = $this->timer->nowInMilliseconds();
-
-        $theoreticalNextIncrementTime = $this->calculateTheoreticalNextIncrementTime(
-            $lastIncrementTimeInMilliseconds,
-            $this->getLastStoredTheoreticalNextIncrementTime($id, $lastIncrementTimeInMilliseconds)
-        );
-
-        $remaining = $this->calculateRemaining($lastIncrementTimeInMilliseconds, $theoreticalNextIncrementTime);
-        $resetAfter = $this->calculateResetAfter($theoreticalNextIncrementTime);
-
-        if ($remaining >= 1) {
-            $this->storeTheoreticalNextIncrementTime($id, $theoreticalNextIncrementTime);
-        }
-
-        return new CounterState($this->limit, $remaining, $resetAfter);
+        $attempts = 0;
+        $isFailStoreUpdatedData = false;
+        do {
+            // Last increment time.
+            // In GCRA it's known as arrival time.
+            $lastIncrementTimeInMilliseconds = $this->timer->nowInMilliseconds();
+
+            $lastStoredTheoreticalNextIncrementTime = $this->getLastStoredTheoreticalNextIncrementTime($id);
+
+            $theoreticalNextIncrementTime = $this->calculateTheoreticalNextIncrementTime(
+                $lastIncrementTimeInMilliseconds,
+                $lastStoredTheoreticalNextIncrementTime
+            );
+
+            $remaining = $this->calculateRemaining($lastIncrementTimeInMilliseconds, $theoreticalNextIncrementTime);
+            $resetAfter = $this->calculateResetAfter($theoreticalNextIncrementTime);
+
+            if ($remaining >= 1) {
```

Comment:
```suggestion
            if ($remaining === 0) {
                break;
```

Reply:
In my opinion, The most common condition should go first. [Should if/else statements be arranged by rareness of cases or difficulty of dealing with them](https://softwareengineering.stackexchange.com/questions/223086/should-if-else-statements-be-arranged-by-rareness-of-cases-or-difficulty-of-deal)

Reply:
What @xepozz suggests here is called "early return" and it increases readability of the code since you don't need `else` in this case. The condition itself is cheap so there's no need to order by frequency.

Reply:
There is no performance-like reasons to keep this style, conditions execute in any way.
But if you like to share articles, [look at this](https://morioh.com/p/d2c5fde959c3).
If you look through Yii3 codebase you may see it almost everywhere.